### PR TITLE
Add examples to docstrings

### PR DESCRIPTION
**Pull request by yuxqiu**
### Goals

- [x] Implement the `__repr__` methods for all the classes
- [x] Add examples to docstrings to illustrate how to use each class
- [x] Investigate `truelearn.quality` (why it produces a very large value)
- [x] Update `truelearn.datasets` implementation (PR was merged into PEEK-Dataset)

### CODE REVIEW

**Comment by KD-7** [comment]
When you run pytest run it via python i.e. python -m pytest (no other args needed as its all in pyproject.toml) this ensures current dir is added to sys path

**Comment by yuxqiu** [comment]
> When you run pytest run it via python i.e. python -m pytest (no other args needed as its all in pyproject.toml) this ensures current dir is added to sys path

The `testpaths` in `pyproject.toml` is wrong. I will fix it in the next commit.

**Comment by KD-7** [comment]
> > When you run pytest run it via python i.e. python -m pytest (no other args needed as its all in pyproject.toml) this ensures current dir is added to sys path
> 
> The `testpaths` in `pyproject.toml` is wrong. I will fix it in the next commit.
The testpaths is correct, you only need to specify the name of the test directory. Doctests are automatically resolved in all files: 
image
As you can see above, this is with testpaths = ['tests'] which it should be

**Comment by yuxqiu** [comment]
> > > When you run pytest run it via python i.e. python -m pytest (no other args needed as its all in pyproject.toml) this ensures current dir is added to sys path
> > 
> > 
> > The `testpaths` in `pyproject.toml` is wrong. I will fix it in the next commit.
> 
> The testpaths is correct, you only need to specify the name of the test directory. Doctests are automatically resolved in all files: image As you can see above, this is with testpaths = ['tests'] which it should be

But when I run `python -m pytest`, I get the following output:

```
=============================================================================================== test session starts ===============================================================================================
platform darwin -- Python 3.7.12, pytest-7.2.1, pluggy-1.0.0 -- /opt/homebrew/Caskroom/miniforge/base/envs/truelearn/bin/python
cachedir: .pytest_cache
rootdir: /Users/.../Documents/OpenSource/TrueLearn-python-library, configfile: pyproject.toml, testpaths: tests
collected 0 items                                                                                                                                                                                                 

============================================================================================== no tests ran in 0.02s ==============================================================================================
```

File: truelearn/learning/_novelty_classifier.py [hunk 1]
```diff
@@ -29,6 +27,40 @@ class NoveltyClassifier(InterestNoveltyKnowledgeBaseClassifier):
...
+        ...     Knowledge({
+        ...         2: KnowledgeComponent(mean=0.37, variance=1e-9),
+        ...         3: KnowledgeComponent(mean=0.41, variance=1e-9),
+        ...     }),
+        ...     Knowledge({
+        ...         1: KnowledgeComponent(mean=0.24, variance=1e-9),
+        ...         3: KnowledgeComponent(mean=0.67, variance=1e-9),
+        ...     }),
+        ... ]
+        >>> events = [EventModel(knowledge) for knowledge in knowledges]
```
Comment (KD-7) [inline truelearn/learning/_novelty_classifier.py:49]: unresolved reference knowledge in EventModel(knowledge)

File: truelearn/learning/_knowledge_classifier.py [hunk 2]
```diff
@@ -35,6 +32,40 @@ class KnowledgeClassifier(InterestNoveltyKnowledgeBaseClassifier):
...
+        ...     Knowledge({
+        ...         2: KnowledgeComponent(mean=0.17, variance=1e-9),
+        ...         3: KnowledgeComponent(mean=0.41, variance=1e-9),
+        ...     }),
+        ...     Knowledge({
+        ...         1: KnowledgeComponent(mean=0.24, variance=1e-9),
+        ...         3: KnowledgeComponent(mean=0.67, variance=1e-9),
+        ...     }),
+        ... ]
+        >>> events = [EventModel(knowledge) for knowledge in knowledges]
```
Comment (KD-7) [inline truelearn/learning/_knowledge_classifier.py:54]: unresolved reference knowledge in Eventmodel(knowledge)

File: truelearn/learning/_interest_classifier.py [hunk 3]
```diff
@@ -43,6 +39,44 @@ class InterestClassifier(InterestNoveltyKnowledgeBaseClassifier):
...
+        ...         3: KnowledgeComponent(mean=0.18, variance=1e-9),
+        ...     }),
+        ...     Knowledge({
+        ...         1: KnowledgeComponent(mean=0.34, variance=1e-9),
+        ...         3: KnowledgeComponent(mean=0.15, variance=1e-9),
+        ...     }),
+        ... ]
+        >>> times = [0, 1024, 5381]
+        >>> events = [
+        ...     EventModel(knowledge, time)
```
Comment (KD-7) [inline truelearn/learning/_interest_classifier.py:63]: unresolved reference time, knowledge in EventModel

File: truelearn/learning/_interest_classifier.py [hunk 4]
```diff
@@ -156,30 +190,43 @@ def __get_decay_func(self) -> Callable[[float], float]:
...
+        event_time (for InterestClassifier).
+
+        Args:
+            learner_kcs:
+                An iterable of learner's knowledge components.
+            content_kcs:
+                An iterable of content's knowledge components.
+            event_time:
+                An optional float representing the event time.
+            y:
```
Comment (KD-7) [inline truelearn/learning/_interest_classifier.py:212]: method has no param y you meant _y

File: truelearn/learning/_ink_classifier.py [hunk 5]
```diff
@@ -24,6 +24,45 @@ class INKClassifier(BaseClassifier):
...
+        ...         3: KnowledgeComponent(mean=0.18, variance=1e-9),
+        ...     }),
+        ...     Knowledge({
+        ...         1: KnowledgeComponent(mean=0.34, variance=1e-9),
+        ...         3: KnowledgeComponent(mean=0.15, variance=1e-9),
+        ...     }),
+        ... ]
+        >>> times = [0, 10514, 53621]
+        >>> events = [
+        ...     EventModel(knowledge, time)
```
Comment (KD-7) [inline truelearn/learning/_ink_classifier.py:48]: unresolved ref knowledge time

**Comment by KD-7** [comment]
There are some additional points i didnt pick up earlier:

- In _peek:
  code duplication 16 lines long starting at 272 and 398
- Shadows line, learner_id, label from out of scope line (line 98 onwards for those vars)

File: truelearn/learning/_novelty_classifier.py (hunk 1 above)
Comment (yuxqiu) [reply truelearn/learning/_novelty_classifier.py:49]: That's weird. Which tool do you run to get this output? This should be a doctest.

**Review by KD-7 — APPROVED** [review]

**Pull request merged**